### ISSUE TITLE
Diff to original impl

### ISSUE THREAD

**Issue by inspirit** [CongWeilin/mtcnn-caffe#1]
Hi, thank you very much for making it available!
I'm curious what is the difference between this and original impl? i see that you dont use ROI regression in 12/24 nets and only apply it in 48net, what is the reason for that?

**Comment by CongWeilin** [comment]
I have tried ROI regression for 12/24 net at the beginning, i change it into 27 classify when i find it's weakness.
1. 12net, 24net can not study well on Euclidean task because it small input. It always disturb the bounding box's shape. 
2. keep bounding box into square before 48net input is important because my training image is square(12*12,24*24,48*48)
3. training data is also different. Original use Wider Face( Rectangle but not square BBox), I used celebA, AFLW with square BBOX.

12net, 24net 27 classify task's reference [paper](https://github.com/CongWeilin/CNN-Cascade-for-FaceDetect/blob/master/cvpr2015_cascnn.pdf).

**Comment by inspirit** [comment]
i see, do you have any ROC numbers/plots to compare to original impl?
can you also please share data preparation code?

**Comment by CongWeilin** [comment]
FDDB result I will upload them in few days, if loaded i will inform you.
Data preparation code is quiet difficult because my data is calculate from different source, the code is really a mass. However, you can refer to [other's](https://github.com/Seanlinx/mtcnn/blob/master/prepare_data/gen_pnet_data.py)

**Comment by inspirit** [comment]
what about detection code? for example how to convert 27 classify result task into offset to calibrate bounding box? this would help a lot to run/evaluate models

**Comment by CongWeilin** [comment]
i just update it in demo. however the caffemodel still need upgrade

**Comment by inspirit** [comment]
great thanx! looking forward for more updates

**Comment by CongWeilin** [comment]
it's my pleasure. i will inform when update

**Comment by inspirit** [comment]
looks like you forgot to commit "48/48net.prototxt" net definition...

**Comment by CongWeilin** [comment]
thanks, now available

**Comment by inspirit** [comment]
hey, what is "12(24)net-only-cls.caffemodel" file for? i see is used to start training. i tried not to provide it for training it Accuracy numbers become very low during train

**Comment by CongWeilin** [comment]
the training process is:
1. train classify task -> `12net-only-cls.caffemodel`
2. finetune regression task on `12net-only-cls.caffemodel`->`12net.caffemodel`

The `runcaffe`'s --weigh is different.

The display take turns between cls_loss and roi_loss. when training classify task, the regression loss will be very high, but it don't backward propagate( some tricks in python loss layer). when training regression task, the acc=0.

I suggest train on your own training data, because the data i upload is just a sample, too big to update all. 

I am going for vacation for two and a half week, sorry for not replying in time.

**Comment by inspirit** [comment]
i see, that make more sense now. 
Can you please provide script example for training 12(24)net-only-cls.caffemodel ? so that i can retrain everything with different data. thank you.

**Comment by CongWeilin** [comment]
[https://github.com/CongWeilin/mtcnn-caffe/tree/master/train-example](https://github.com/CongWeilin/mtcnn-caffe/tree/master/train-example) provide an example on how to train 12net

**Comment by inspirit** [comment]
thank you and have a nice holiday.

**Comment by inspirit** [comment]
one more quick question: i see you normalize image data to -127/+127 any specific reason not to scale it to -1/+1 ? i also noticed that if I try to scale it to -1/+1 training scores getting small again

**Comment by inspirit** [comment]
oh i believe you cant simply flip image for ROI train:
```im = cv2.flip(cur_data[0], random.choice([-1,0,1]))``` because you also need to change label for that

**Comment by CongWeilin** [comment]
flip only applied on cls_task.

**Comment by inspirit** [comment]
yup sorry flip is only for cls -- its OK. any comments on data scale to -127/+127 ?

**Comment by CongWeilin** [comment]
this can make convergence faster and more acc . Refer to [Batch Normalization Accelerating Deep Network Training by Reducing Internal Covariate Shift](https://arxiv.org/abs/1502.03167). This operation can separate date from [0,+) to (-,+).

**Comment by inspirit** [comment]
thank you for the info.

**Comment by inspirit** [comment]
did you try implementing online-negative-mining?

**Comment by CongWeilin** [comment]
no, OHEM is used for training data is not that much big. Refer to faster-rcnn's writer RBG's papaer[https://arxiv.org/pdf/1604.03540.pdf](https://arxiv.org/pdf/1604.03540.pdf)

**Comment by CongWeilin** [comment]
If having new questions, please open another issue so that others can check it easily o(^▽^)o

**Comment by buweiaini** [comment]
Dear CongWeiling :

Could you please share me an example of training? I can't get access to the website below
"https://github.com/CongWeilin/mtcnn-caffe/tree/master/train-example"

Thank you!

**Comment by CongWeilin** [comment]
You just need to get into 12net and run `runcaffe`. The train-example is useless since the model have update to be more acc. You can also check out Q&A to solve your questions

**Comment by buweiaini** [comment]
another question :   I find that you use the layer type "Python",,,,
so do I have to modify the C++ source code of caffe for using "Python" layer?

**Comment by CongWeilin** [comment]
no. You should edit `caffe/Makefile.config` to support python-layer and re-make

**Comment by buweiaini** [comment]
thanks bro!

**Comment by bpf476499866** [comment]
@CongWeilin 
Hello and sorry to bother you! 
——I want to know why you do y-flip (and x-y-flip) when training cls-task, in pythonlayer.py, function  load_next_image, line 163-164:【if random.choice([0,1])==1:  im = cv2.flip(im,random.choice([-1,0,1]))】;
——I understand to flip or to rotate an image can add data and make classifier robuster,
——But I think to make the image up-side-down does no good to train a face classifier, unless the classifier will meet many reversed faces……
